### PR TITLE
WIP: create dofs from interpolations

### PR DESCRIPTION
**Pull request by KristofferC**
Try for example:

```jl
using JuAFEM

g = generate_grid(Quadrilateral, (2,1))
dh = DofHandler(g)
ip = Lagrange{2, RefCube, 3}()
push!(dh, :u, ip, 2)
push!(dh, :p, ip, 1)

free_dof = 1
storages = []
for i in 1:length(dh.interpolations)
  ip = dh.interpolations[i]
  storage = JuAFEM.DofStorage(g, ip, dh.dof_dims[i])
  push!(storages, storage)

  free_dof = JuAFEM.add_node_dofs(g, ip, storage, free_dof)
  free_dof = JuAFEM.add_edge_dofs(g, ip, storage, free_dof)
  free_dof = JuAFEM.add_cell_dofs(g, ip, storage, free_dof)
end
```

Should output something like:

```
Adding node dofs: (1, 2) for global vertex: 1 local vertex: 1
Adding node dofs: (3, 4) for global vertex: 2 local vertex: 2
Adding node dofs: (5, 6) for global vertex: 5 local vertex: 3
Adding node dofs: (7, 8) for global vertex: 4 local vertex: 4
Reusing node dofs (3, 4) for global vertex: 2 local vertex: 1
Adding node dofs: (9, 10) for global vertex: 3 local vertex: 2
Adding node dofs: (11, 12) for global vertex: 6 local vertex: 3
Reusing node dofs (5, 6) for global vertex: 5 local vertex: 4
Adding edge dofs: (13, 14, 15, 16), for global edge (1, 2)  local edge: 1, orientation: true
Adding edge dofs: (17, 18, 19, 20), for global edge (2, 5)  local edge: 2, orientation: true
Adding edge dofs: (21, 22, 23, 24), for global edge (4, 5)  local edge: 3, orientation: false
Adding edge dofs: (25, 26, 27, 28), for global edge (1, 4)  local edge: 4, orientation: false
Adding edge dofs: (29, 30, 31, 32), for global edge (2, 3)  local edge: 1, orientation: true
Adding edge dofs: (33, 34, 35, 36), for global edge (3, 6)  local edge: 2, orientation: true
Adding edge dofs: (37, 38, 39, 40), for global edge (5, 6)  local edge: 3, orientation: false
Reusing edge dofs: (17, 18, 19, 20), for global edge (2, 5) local edge: 4, same orientation: false
Adding cell dofs: [41, 42, 43, 44, 45, 46, 47, 48], for global cell 1
Adding cell dofs: [49, 50, 51, 52, 53, 54, 55, 56], for global cell 2
Adding node dofs: (57,) for global vertex: 1 local vertex: 1
Adding node dofs: (58,) for global vertex: 2 local vertex: 2
Adding node dofs: (59,) for global vertex: 5 local vertex: 3
Adding node dofs: (60,) for global vertex: 4 local vertex: 4
Reusing node dofs (58,) for global vertex: 2 local vertex: 1
Adding node dofs: (61,) for global vertex: 3 local vertex: 2
Adding node dofs: (62,) for global vertex: 6 local vertex: 3
Reusing node dofs (59,) for global vertex: 5 local vertex: 4
Adding edge dofs: (63, 64), for global edge (1, 2)  local edge: 1, orientation: true
Adding edge dofs: (65, 66), for global edge (2, 5)  local edge: 2, orientation: true
Adding edge dofs: (67, 68), for global edge (4, 5)  local edge: 3, orientation: false
Adding edge dofs: (69, 70), for global edge (1, 4)  local edge: 4, orientation: false
Adding edge dofs: (71, 72), for global edge (2, 3)  local edge: 1, orientation: true
Adding edge dofs: (73, 74), for global edge (3, 6)  local edge: 2, orientation: true
Adding edge dofs: (75, 76), for global edge (5, 6)  local edge: 3, orientation: false
Reusing edge dofs: (65, 66), for global edge (2, 5) local edge: 4, same orientation: false
Adding cell dofs: [77, 78, 79, 80], for global cell 1
Adding cell dofs: [81, 82, 83, 84], for global cell 2
```

### CODE REVIEW

**Comment by KristofferC** [comment]
More wip.

Can be tried with:

```jl
g = generate_grid(Quadrilateral, (3,3))
dh = DofHandler(g)
ip = Lagrange{2, RefCube, 2}()
push!(dh, :u, ip, 2)
push!(dh, :p, ip, 1)

close!(dh)
dh.dofs_cells
```

File: src/Dofs/DofHandler.jl [hunk 1]
```diff
@@ -76,26 +78,73 @@ julia> vtk_point_data(vtkfile
...
 type DofHandler{dim, N, T, M}
-    dofs_nodes::Matrix{Int}
-    dofs_cells::Matrix{Int} # TODO <- Is this needed or just extract from dofs_nodes?
+    dofs_cells::Matrix{Int}
     field_names::Vector{Symbol}
+    interpolations::Vector{Interpolation}
     dof_dims::Vector{Int}
-    closed::ScalarWrapper{Bool}
-    dofs_vec::Vector{Int}
+    closed::JuAFEM.ScalarWrapper{Bool}
```
Comment (fredrikekre) [inline src/Dofs/DofHandler.jl]: why the `JuAFEM.`?

File: src/Dofs/DofHandler.jl [hunk 2]
```diff
@@ -76,26 +78,73 @@ julia> vtk_point_data(vtkfile
...
     grid::Grid{dim, N, T, M}
 end
 
-function DofHandler(m::Grid)
-    DofHandler(Matrix{Int}(0, 0), Matrix{Int}(0, 0), Symbol[], Int[], ScalarWrapper(false), Int[], m)
+function DofHandler(grid::Grid)
+    DofHandler(Matrix{Int}(0, 0), Symbol[], Interpolation[], Int[], JuAFEM.ScalarWrapper(false), grid)
```
Comment (fredrikekre) [inline src/Dofs/DofHandler.jl]: why the `JuAFEM.`?

File: src/Dofs/DofHandler.jl [hunk 3]
```diff
@@ -76,26 +78,73 @@ julia> vtk_point_data(vtkfile
...
+end
+
+isclosed(dh::DofHandler) = dh.closed[]
+
+
+# Add a field to the dofhandler ex `push!(dh, :u, 3)`
+function Base.push!(dh::DofHandler, name::Symbol, interpolation::Interpolation, dim::Int)
```
Comment (fredrikekre) [inline src/Dofs/DofHandler.jl]: This need to check that the interpolation is compatible with the cell in the grid (same dim and shape at least)

File: src/Dofs/DofHandler.jl [hunk 4]
```diff
@@ -104,99 +153,204 @@ function Base.show(io::IO, dh::DofHandler)
         print(io, "  Dofs per cell: ", ndofs_per_cell(dh))
     end
 end
+immutable SortedNTuple{N, T}
+  n::NTuple{N, T}
```
Comment (fredrikekre) [inline src/Dofs/DofHandler.jl]: 4 space indent (kappa)

File: src/Dofs/DofHandler.jl [hunk 5]
```diff
@@ -104,99 +153,204 @@ function Base.show(io::IO, dh::DofHandler)
...
+    dof_numbers::NTuple{N, Int} # Number of the dofs
 end
 
-# Add a collection of fields
-function Base.push!(dh::DofHandler, names::Vector{Symbol}, dims)
-    @assert length(names) == length(dims)
-    for i in 1:length(names)
-        push!(dh, names[i], dims[i])
-    end
+immutable GlobalEdge{N}
```
Comment (fredrikekre) [inline src/Dofs/DofHandler.jl]: I assume there will be a `GlobalFace` too later on?

File: src/Dofs/DofHandler.jl [hunk 6]
```diff
@@ -104,99 +153,204 @@ function Base.show(io::IO, dh::DofHandler)
...
+    sign::Bool                  # "sign" of the edge
```
Comment (fredrikekre) [inline src/Dofs/DofHandler.jl]: isn't a global edge always positive?

File: src/Dofs/DofHandler.jl [hunk 7]
```diff
@@ -104,99 +153,204 @@ function Base.show(io::IO, dh::DofHandler)
...
-    i = 0
-    for name in dh.field_names
-        i += 1
-        if name == field_name
-            return dh.dof_dims[i]
-        end
+function DofStorage(grid, ip::Interpolation, dim)
+  N = dim
+  M = get_n_edgedofs(ip) * dim
+  cell_vertex_dofs = zeros(Int, dim,                  get_n_vertices(ip), JuAFEM.getncells(grid))
```
Comment (fredrikekre) [inline src/Dofs/DofHandler.jl]: `JuAFEM.` ?

Comment (KristofferC) [inline src/Dofs/DofHandler.jl]: Because I developed it in a separate file.

Comment (fredrikekre) [inline src/Dofs/DofHandler.jl]: okay

File: src/Dofs/DofHandler.jl (hunk 6 above)
Comment (KristofferC) [inline src/Dofs/DofHandler.jl]: I guess. This is whether the cell that created dofs for this edge had a positive or negative edge.

**Comment by KristofferC** [comment]
Played a bit with plot recipes: https://github.com/KristofferC/JuAFEM.jl/blob/kc/dof_builder/examples/interpolations/lagrange.ipynb

**Comment by fredrikekre** [comment]
Thats awesome!

**Comment by KristofferC** [comment]
FWIW, this stuff is probably the hardest stuff I've ever programmed... xD

**Comment by KristofferC** [comment]
Simple 1D heat: https://github.com/KristofferC/JuAFEM.jl/blob/kc/dof_builder/examples/heat_1D.ipynb

**Comment by KristofferC** [comment]
Updated plots with shape functions

**Comment by KristofferC** [comment]
3d baby http://nbviewer.jupyter.org/github/KristofferC/JuAFEM.jl/blob/0ffb72a37dd9bc9139da65ffa6912bf9edcdc6b9/examples/interpolations/lagrange.ipynb

**Comment by KristofferC** [comment]
Note to self: It's annoying to store dof numbers in immutable vectors because when doing dof renumbering, these are hard to change.

**Comment by fredrikekre** [comment]
Sweet!
But dude... produce smaller PR's! :stuck_out_tongue:

**Comment by KristofferC** [comment]
I'll try to break out the stuff from here which can be applied cleanly to master. Also, this PR is still very "exploratory"...

**Pull request closed without merging**